shap/utils/_clustering.py

```diff
@@ -228,37 +253,31 @@ def hclust(
 
     # build the distance matrix
     if metric == "xgboost_distances_r2":
-        dist_full = xgboost_distances_r2(X, y, random_state=random_state)
+        dist_full: np.ndarray = xgboost_distances_r2(X_arr, y, random_state=random_state)
 
         # build a condensed upper triangular version by taking the max distance from either direction
-        dist_list = []
-        for i in range(dist_full.shape[0]):
-            for j in range(i + 1, dist_full.shape[1]):
-                if i != j:
-                    if linkage == "single":
-                        dist_list.append(min(dist_full[i, j], dist_full[j, i]))
-                    elif linkage == "complete":
-                        dist_list.append(max(dist_full[i, j], dist_full[j, i]))
-                    elif linkage == "average":
-                        dist_list.append((dist_full[i, j] + dist_full[j, i]) / 2)
-                    else:
-                        raise ValueError("Unsupported linkage type!")
+        dist_list: list[float] = []
+        for i, j in it.combinations_with_replacement(range(len(dist_full)), 2):
+            if i == j:
+                continue
```

Comment:
This `it.combinations_with_replacement` is identical (element-wise) with the nested for loop, it just saves us one level of indentation.

Also note here that `dist_full` is necessarily a square matrix. So `shape[1] == shape[0]`.

![image](https://github.com/user-attachments/assets/c9f5b209-06c3-4578-a27e-78c7eee8104b)

Reply:
@thatlittleboy if I understand correctly, is this logic:

```python
for i, j in it.combinations_with_replacement(range(len(dist_full)), 2):
    if i == j:
        continue
```

equivalent to just:

```python
for i, j in it.combinations(range(len(dist_full)), 2):
```

?

Reply:
Thanks @connortann , nice catch. raised a separate PR to fix this. #3776